File: updated_code/code/java/1796-second-largest-digit-in-a-string.java
package com. improvedCode;

import java.util.*;

public class Solution {
    /**
     * Given an alphanumeric string `s`, return _the **second largest** numerical
     * digit that appears in _`s` _, or_`-1` _if it does not exist_.
     * 
     * @param s the input string
     * @return the second largest numerical digit, or -1 if it does not exist
     */
    public int secondHighest(String s) {
        Set<Integer> set = new HashSet<>();
        for (char c : s.toCharArray()) {
            if (Character.isDigit(c)) {
                set.add(Integer.parseInt(String.valueOf(c)));
            }
        }
        if (set.size() < 2) {
            return -1;
        }
        int secondHighest = -1;
        for (int num : set) {
            if (num > secondHighest && num != Collections.max(set)) {
                secondHighest = num;
            }
        }
        return secondHighest;
    }
}

File: updated_code/code/python/1796-second-largest-digit-in-a-string.py
"""
Description:
Given an alphanumeric string s, return the **second largest** numerical digit that appears in s, or -1 if it does not exist.

Example 1:
Input: s = "dfa12321afd"    Output: 2    Explanation: The digits that appear in s are [1, 2, 3]. The second largest digit is 2.

Example 2:
Input: s = "abc1111"    Output: -1    Explanation: The digits that appear in s are [1]. There is no second largest digit.

Constraints:
* 1 <= s.length <= 500
* s consists of only lowercase English letters and/or digits.
"""
class SecondHighestDigitSolutions(object):
    """
    This class contains a method to find the second highest numerical digit in a string.
    """
    def second_highest(self, s: str) -> int:
        """
        This method returns the second highest numerical digit in a string, or -1 if it does not exist.
        
        Args:
        s (str): An alphanumeric string.
        
        Returns:
        int: The second highest numerical digit in the string, or -1 if it does not exist.
        """
        # Convert the string to a set of digits to remove duplicates
        digits = set(c for c in s if c.isdigit())
        
        # Convert the set of digits to a list and sort it in descending order
        sorted_digits = sorted(map(int, digits), reverse=True)
        
        # If the list of digits contains less than 2 elements, return -1
        if len(sorted_digits) < 2:
            return -1
        else:
            # Otherwise, return the second highest digit
            return sorted_digits[1]